### PR TITLE
feat: lock versions in docker file

### PR DESCRIPTION
**Pull request by rti**
This change locks the versions of
 - ollama (by taking ollama from the upstream docker image)
 - python libraries (by installing them via requirements.txt)

This also reorders the Dockerfile for faster rebuilds when developing
fronend or backend.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -42,24 +47,31 @@ RUN ollama serve & while ! curl http://localhost:11434; do sleep 1; done; ollama
...
-RUN cd frontend && yarn install && yarn build
+# Install frontend dependencies
+COPY --chmod=755 frontend/package.json frontend/package.json
+COPY --chmod=755 frontend/yarn.lock frontend/yarn.lock
+RUN cd frontend && yarn install
+
+
+# Copy and build frontend for production (into the frontend/dist folder)
+COPY --chmod=755 frontend frontend
+RUN cd frontend && yarn build
```
Comment (exowanderer) [inline Dockerfile:67]: I am curious why the Dockerfile changes directory to frontend twice

**Review by exowanderer — COMMENTED** [review]
The Dockerfile and requirements.txt file are well designed. I like this inclusion of the requirements.txt from outside the Dockerfile.

Because of the complications with Ubuntu 20.04 and Linux Kernel v5, I added the following to my own version of your Dockerfile, which is also reflected in PR#20 

```Dockerfile
# Install node from upstream, ubuntu packages are too old

...
...

RUN npm cache clean -f
RUN npm install -g n
RUN n 18.17.1

...
...
RUN npm install -g yarn
```

File: Dockerfile (hunk 1 above)
Comment (exowanderer) [reply Dockerfile:67]: Actually, I commented out both of these lines and kept the original towards the end after installing node18 and, thus, yarn.

```Dockerfile
# RUN cd frontend && yarn install
...
# RUN cd frontend && yarn build
...
RUN cd frontend && yarn install && yarn build
...
# Container startup script
COPY --chmod=755 start.sh /start.sh
CMD [ "/start.sh" ]
```

Comment (rti) [reply Dockerfile:67]: > I am curious why the Dockerfile changes directory to frontend twice

Docker build is weird, each `RUN` is happening in a new container, so everything from the last `RUN` (processes, working directory) is lost. That's why every `RUN` changes the directory individually. What would work though is `WORKDIR`, this remains set throughout the `Dockerfile` including in the resulting container. I think it is a matter of taste what to use in the build process.

Comment (rti) [reply Dockerfile:67]: > Actually, I commented out both of these lines and kept the original towards the end after installing node18 and, thus, yarn. (See below)

I split `yarn install` and `yarn build` intentionally into two `RUN`s in order to use the docker build cache effectively. `yarn install` is only repeated when dependencies in `package.json` / `yarn.lock` change, `yarn build` is repeated when frontend source files change.

Comment (rti) [reply Dockerfile:67]: @exowanderer Could you please see if #22 also builds on your machine? This is basically this branch, but downgraded to node 18.

**Comment by rti** [comment]
follow up in #22

**Pull request closed without merging**